File: apps/sonar/composeApp/src/commonMain/kotlin/chat/bitchat/sonar/screens/SonarChannelScreen.kt
package chat.bitchat.sonar.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.clickable
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import chat.bitchat.sonar.Screen
import chat.bitchat.sonar.SlashHints
import chat.bitchat.sonar.SonarAppState
import chat.bitchat.sonar.SonarChannelMsg
import chat.bitchat.sonar.ToastBar
import chat.bitchat.sonar.ui.SNBanner
import chat.bitchat.sonar.ui.SNBannerTone
import chat.bitchat.sonar.ui.SNDot
import chat.bitchat.sonar.ui.SNEmptyState
import chat.bitchat.sonar.ui.SNIcon
import chat.bitchat.sonar.ui.SNIconButton
import chat.bitchat.sonar.ui.SNIconName
import chat.bitchat.sonar.ui.SNPrimaryButton
import chat.bitchat.sonar.ui.SNGhostButton
import chat.bitchat.sonar.ui.authorColor
import chat.bitchat.sonar.ui.sonar

@Composable
fun SonarChannelScreen(state: SonarAppState, screen: Screen.Channel) {
    val s = sonar
    var draft by remember { mutableStateOf("") }
    var authorSheet by remember { mutableStateOf<SonarChannelMsg?>(null) }
    val listState = rememberLazyListState()
    LaunchedEffect(state.channelMsgs.size) {
        if (state.channelMsgs.isNotEmpty()) listState.animateScrollToItem(state.channelMsgs.size - 1)
    }
    // Humanize the channel: the design NEVER shows a raw geohash as a label — use
    // the location channel's name + tier + live "N here now" count (state.presence).
    val gc = state.locationChannels.firstOrNull { it.geohash == screen.geohash }
    val isMesh = screen.geohash == "mesh"
    val name = gc?.name ?: if (isMesh) "Mesh" else "#${screen.geohash}"
    val here = state.presence(screen.geohash)
    val tier = gc?.level?.label ?: if (isMesh) "Bluetooth range" else "channel"
    val headerSub = if (here > 0) "Public · $here here now" else "Public · $tier"

    Column(Modifier.fillMaxSize().background(s.bg)) {
        // header: place tile + name + status
        Row(
            Modifier.fillMaxWidth().padding(start = 6.dp, end = 16.dp, top = 12.dp, bottom = 8.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            SNIconButton(SNIconName.Back, onClick = { state.back() })
            Box(
                Modifier.size(36.dp).clip(RoundedCornerShape(11.dp)).background(s.accentSoft),
                contentAlignment = Alignment.Center
            ) { SNIcon(SNIconName.Pin, 18.dp, s.accentDeep) }
            Spacer(Modifier.width(10.dp))
            Column(Modifier.weight(1f)) {
                Text(name, color = s.text, fontSize = 16.sp, fontWeight = FontWeight.Bold, maxLines = 1)
                Row(verticalAlignment = Alignment.CenterVertically) {
                    SNDot(s.green, 6.dp)
                    Spacer(Modifier.width(5.dp))
                    Text(headerSub, color = s.text2, fontSize = 11.5.sp)
                }
            }
            // Trailing: pin/save this channel to home (geohash channels only — Mesh
            // is always present), then people-nearby (radar) per the design.
            if (!isMesh) {
                val saved = state.isSaved(screen.geohash)
                SNIconButton(
                    if (saved) SNIconName.BookmarkFill else SNIconName.Bookmark,
                    tint = if (saved) s.accent else s.text2,
                    onClick = { state.toggleSaved(screen.geohash) }
                )
            }
            SNIconButton(SNIconName.Rings, onClick = { state.push(Screen.Nearby) })
        }

        SNBanner(
            icon = SNIconName.People, tone = SNBannerTone.Public,
            bold = "Public channel", rest = " — anyone nearby can read"
        )

        if (state.channelMsgs.isEmpty()) {
            Box(Modifier.weight(1f).fillMaxWidth()) {
                if (isMesh) {
                    SNEmptyState(
                        icon = SNIconName.Mesh, iconSize = 26.dp,
                        title = "Bluetooth mesh",
                        desc = "Messages here reach everyone in Bluetooth range — no internet needed. Say hi."
                    )
                } else {
                    SNEmptyState(
                        icon = SNIconName.Pin, iconSize = 26.dp,
                        title = "Quiet in $name right now",
                        desc = if (here > 0) "$here ${if (here == 1) "person is" else "people are"} in range of this channel today. Say hi."
                               else "Be the first to say something in $name."
                    )
                }
            }
        } else {
            LazyColumn(
                Modifier.weight(1f).fillMaxWidth(),
                state = listState,
                contentPadding = PaddingValues(horizontal = 14.dp, vertical = 8.dp)
            ) {
                items(state.channelMsgs, key = { it.id }) { m ->
                    ChannelBubble(m) {
                        if (!m.mine) authorSheet = m
                    }
                }
            }
        }

        if (draft.startsWith("/")) SlashHints(draft) { draft = it }

        // composer
        Row(Modifier.fillMaxWidth().padding(10.dp), verticalAlignment = Alignment.Bottom) {
            Box(
                Modifier.weight(1f).clip(RoundedCornerShape(22.dp)).background(s.surface2)
                    .heightIn(min = 46.dp)
                    .padding(horizontal = 16.dp, vertical = 12.dp)
            ) {
                if (draft.isEmpty()) Text("Message $name", color = s.text3, fontSize = 16.sp)
                BasicTextField(
                    value = draft, onValueChange = { draft = it },
                    textStyle = TextStyle(color = s.text, fontSize = 16.sp),
                    cursorBrush = SolidColor(s.accent),
                    singleLine = false,
                    maxLines = 5,
                    modifier = Modifier.fillMaxWidth()
                )
            }
            Spacer(Modifier.width(8.dp))
            Box(
                Modifier.size(46.dp).clip(CircleShape).background(s.netFill)
                    .clickable {
                        val d = draft
                        draft = ""
                        if (!state.handleCommand(d, name, channelGeohash = screen.geohash, chatId = null)) {
                            state.sendChannelMsg(screen.geohash, d)
                        }
                    },
                contentAlignment = Alignment.Center
            ) { Text("↑", color = s.onNet, fontSize = 20.sp, fontWeight = FontWeight.Bold) }
        }
    }
    authorSheet?.let { author ->
        ChannelAuthorSheet(
            author = author,
            blocked = state.isChannelAuthorBlocked(author.senderPubkey),
            onPrivateChat = {
                authorSheet = null
                state.openGeoDm(screen.geohash, author.senderPubkey, author.author)
            },
            onBlock = {
                authorSheet = null
                state.setChannelAuthorBlocked(author.senderPubkey, author.author, blocked = true)
            },
            onClose = { authorSheet = null },
        )
    }
    state.toast?.let { ToastBar(it) { state.toast = null } }
}

@Composable
private fun ChannelAuthorSheet(
    author: SonarChannelMsg,
    blocked: Boolean,
    onPrivateChat: () -> Unit,
    onBlock: () -> Unit,
    onClose: () -> Unit,
) {
    val s = sonar
    Box(
        Modifier.fillMaxSize().background(s.scrim).clickable(onClick = onClose),
        contentAlignment = Alignment.BottomCenter
    ) {
        Surface(color = s.surface, shape = RoundedCornerShape(topStart = 22.dp, topEnd = 22.dp)) {
            Column(Modifier.fillMaxWidth().padding(20.dp), horizontalAlignment = Alignment.CenterHorizontally) {
                Text(author.author, color = s.text, fontSize = 18.sp, fontWeight = FontWeight.Bold, maxLines = 1, overflow = TextOverflow.Ellipsis)
                Spacer(Modifier.heightIn(min = 14.dp))
                SNPrimaryButton("Open private chat", disabled = blocked) { onPrivateChat() }
                Spacer(Modifier.heightIn(min = 8.dp))
                SNGhostButton("Block author") { onBlock() }
            }
        }
    }
}

@Composable
private fun ChannelBubble(m: SonarChannelMsg, onTapAuthor: () -> Unit) {
    val s = sonar
    Column(
        Modifier.fillMaxWidth().padding(vertical = 3.dp),
        horizontalAlignment = if (m.mine) Alignment.End else Alignment.Start
    ) {
        if (!m.mine) {
            Text(
                m.author, color = authorColor(m.author, s.isDark),
                fontSize = 12.sp, fontWeight = FontWeight.Bold,
                modifier = Modifier.padding(start = 12.dp, bottom = 3.dp).clickable(onClick = onTapAuthor)
            )
        }
        Box(
            Modifier.clip(RoundedCornerShape(18.dp))
                .background(if (m.mine) s.netFill else s.bubbleOther)
                .padding(horizontal = 12.dp, vertical = 8.dp)
        ) {
            Text(m.content, color = if (m.mine) s.onNet else s.text, fontSize = 16.sp)
        }
    }
}

File: ios/bitchat/Views/Sonar/SonarChannelScreen.swift
//
// SonarChannelScreen.swift
// bitchat
//
// Public location-channel screen (ChannelScreen in
// design/handoff/project/sonar/screens.jsx), backed by the real timeline:
// ChatViewModel messages for the selected #mesh / geohash channel, sent
// through the existing send path (mesh broadcast or kind-20000 over Nostr).
//
// This is free and unencumbered software released into the public domain.
// For more information, see <https://unlicense.org>
//

import SwiftUI

struct SonarChannelScreen: View {
    @EnvironmentObject private var store: SonarAppStore
    // NB: GeohashBookmarksStore.shared is LocationStateManager, whose
    // objectWillChange the store already republishes — so toggling the bookmark
    // re-renders this view through `store`. No direct @ObservedObject needed.
    let chId: String

    @State private var sheet = false
    @State private var authorSheet: SonarAppStore.SNChannelAuthor?
    @State private var toast: String?
    @State private var previewPackCoordinate: String?

    private var ch: SNChannelItem { store.channelItem(chId) }

    /// Raw geohash for a location channel ("geo:<gh>"), nil for the Mesh channel
    /// (Mesh is always present and never bookmarkable).
    private var geohash: String? {
        chId.hasPrefix("geo:") ? String(chId.dropFirst(4)) : nil
    }

    /// Channel routing is real: geohash channels go over Nostr (kind 20000),
    /// the #mesh channel broadcasts over Bluetooth.
    private var transport: SNVia {
        chId == "mesh" ? .mesh : .internet
    }

    /// Target for /slap: most recent other participant, else the whole channel.
    private var slapTarget: String {
        let msgs = store.chMsgs(chId)
        return msgs.last(where: { !$0.mine && !$0.action })?.author ?? "everyone"
    }

    var body: some View {
        VStack(spacing: 0) {
            SNNavHeader(onBack: { store.pop() }) {
                SNPlaceTile(size: 36, icon: chId == "mesh" ? .mesh : .pin)
                SNHeaderTitle(name: ch.name) {
                    SNDot(color: SonarTheme.green, small: true)
                    Text(verbatim: ch.sub)
                }
            } trailing: {
                HStack(spacing: 2) {
                    // Save/unsave this channel to the home "Saved channels" list
                    // (geohash channels only — Mesh is always present).
                    if let gh = geohash {
                        let saved = GeohashBookmarksStore.shared.isBookmarked(gh)
                        SNIconButton(action: {
                            GeohashBookmarksStore.shared.toggle(gh)
                            showToast(GeohashBookmarksStore.shared.isBookmarked(gh) ? "Channel saved" : "Removed from saved channels")
                        }) {
                            Image(systemName: saved ? "bookmark.fill" : "bookmark")
                                .font(.system(size: 18, weight: .semibold))
                                .foregroundColor(saved ? SonarTheme.accent : SonarTheme.text2)
                        }
                        .accessibilityLabel(saved ? "Unsave channel" : "Save channel")
                    }
                    SNIconButton(action: { store.push(.nearby) }) {
                        SNIcon(name: .rings, size: 20)
                    }
                    .accessibilityLabel("People nearby")
                }
            }

            SNBanner(icon: .people, tone: .publicRoom, bold: "Public channel", rest: " — anyone nearby can read")

            let msgs = store.chMsgs(chId)
            if msgs.isEmpty {
                SNEmptyState(
                    icon: .pin,
                    iconSize: 26,
                    amber: true,
                    title: "Quiet in \(ch.name) right now",
                    desc: ch.count > 0
                        ? "\(ch.count) people are in range of this channel. Say hi."
                        : "Nobody has said anything yet. Say hi."
                )
            } else {
                SNMsgList(msgs: msgs, showAuthors: true, onTapAuthor: { m in
                    guard !m.mine else { return }
                    if let author = store.channelAuthor(forMessage: m.id) {
                        authorSheet = author
                    } else {
                        // Their per-location identity has left the channel, so
                        // we can't open a DM right now (honest "offline" signal).
                        showToast("\(m.author ?? "Questa persona") non \u{00E8} pi\u{00F9} nel canale")
                    }
                }, loadSticker: { await store.stickerImageData(for: $0, userInitiated: $1) },
                    onTapPack: { previewPackCoordinate = $0 })
            }

            SNComposer(
                placeholder: "Message \(ch.name)",
                transport: transport,
                onSend: { store.sendCh(chId, $0) },
                onPlus: { sheet = true },
                onCommand: { cmd in
                    store.onCommand(.init(type: .ch, id: chId, target: slapTarget), cmd)
                },
                onSticker: { sticker, coord in
                    if !store.sendStickerToChannel(chId, sticker: sticker, packCoordinate: coord) {
                        showToast("Stickers aren't supported in public channels yet")
                    }
                },
                loadStickerPack: { author, identifier, relays in
                    await store.stickerPack(authorPubkeyHex: author, identifier: identifier, relayUrls: relays)
                },
                loadStickerImage: { await store.stickerImageData(url: $0, expectedSha256: $1) },
                fetchInstalledPacks: { await store.fetchInstalledPacks() },
                cachedStickerPacks: { store.cachedStickerPacks() },
                voiceEnabled: false
            )
        }
        .background(SonarTheme.bg.ignoresSafeArea())
        .overlay {
            if let coord = previewPackCoordinate {
                StickerPackPreviewSheet(
                    coordinate: coord,
                    loadPack: { author, identifier, relays in
                        await store.stickerPack(authorPubkeyHex: author, identifier: identifier, relayUrls: relays)
                    },
                    loadImage: { await store.stickerImageData(url: $0, expectedSha256: $1) },
                    installPack: { await store.installStickerPack(coordinate: $0) },
                    uninstallPack: { await store.uninstallStickerPack(coordinate: $0) },
                    isInstalled: { packCoord in
                        let installed = await store.fetchInstalledPacks()
                        return snStickerPackInstalledState(
                            coordinate: packCoord,
                            refreshedCoordinates: installed,
                            cachedInstalled: store.isStickerPackInstalled(packCoord)
                        )
                    },
                    onClose: { previewPackCoordinate = nil }
                )
            }
        }
        .overlay(alignment: .bottom) { toastView }
        .animation(.easeOut(duration: 0.2), value: toast)
        .onAppear { store.ensureChannelSelected(chId) }
        .snSheet(isPresented: $sheet, title: "Add to your message") {
            VStack(spacing: 0) {
                SNActionRow(icon: .people, label: "People nearby", desc: "See who can hear you over Bluetooth") {
                    sheet = false
                    store.push(.nearby)
                }
            }
        }
        .snSheet(
            isPresented: Binding(
                get: { authorSheet != nil },
                set: { if !$0 { authorSheet = nil } }
            ),
            title: authorSheet.map { "Message \($0.name)" } ?? ""
        ) {
            if let author = authorSheet {
                VStack(spacing: 0) {
                    SNActionRow(
                        icon: .lock, label: "Open private chat",
                        desc: "End-to-end encrypted, over the internet"
                    ) {
                        authorSheet = nil
                        store.openChannelDM(author)
                    }
                    SNActionRow(
                        icon: .trash, label: "Block \(author.name)",
                        desc: "You won\u{2019}t see their messages anymore"
                    ) {
                        let name = author.name
                        authorSheet = nil
                        store.blockChannelAuthor(author)
                        showToast("\(name) blocked")
                    }
                }
            }
        }
    }

    // sn-toast: transient bottom snackbar in the app style (auto-dismisses).
    @ViewBuilder
    private var toastView: some View {
        if let toast {
            Text(verbatim: toast)
                .font(SonarTheme.uiFont(size: 13.5, weight: .medium))
                .foregroundColor(SonarTheme.text)
                .multilineTextAlignment(.center)
                .padding(EdgeInsets(top: 11, leading: 16, bottom: 11, trailing: 16))
                .background(
                    RoundedRectangle(cornerRadius: 13, style: .continuous)
                        .fill(SonarTheme.surface2)
                        .shadow(color: Color.black.opacity(0.18), radius: 12, y: 6)
                )
                .overlay(
                    RoundedRectangle(cornerRadius: 13, style: .continuous)
                        .strokeBorder(SonarTheme.hairline, lineWidth: 1)
                )
                .padding(.horizontal, 24)
                .padding(.bottom, 88)
                .transition(.move(edge: .bottom).combined(with: .opacity))
        }
    }

    private func showToast(_ text: String) {
        toast = text
        DispatchQueue.main.asyncAfter(deadline: .now() + 2.6) {
            if toast == text { toast = nil }
        }
    }
}
